compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -1938,10 +1948,48 @@ public void visit(BLangRecordDestructure recordDeStmt) {
             return;
         }
         typeChecker.checkExpr(recordDeStmt.expr, this.env);
+        if (checkValidMappingBindingPatternForRecordVarRef(recordDeStmt.varRef.recordRefFields,
+                recordDeStmt.expr.getBType())) {
+            dlog.error(recordDeStmt.expr.pos, DiagnosticErrorCode.INVALID_OPTIONAL_FIELD_IN_MAPPING_BINDING_PATTERN);
+        }
         checkRecordVarRefEquivalency(recordDeStmt.pos, recordDeStmt.varRef, recordDeStmt.expr.getBType(),
                                      recordDeStmt.expr.pos);
     }
 
+    private boolean checkValidMappingBindingPatternForRecordVarRef(List<BLangRecordVarRefKeyValue> recordRefFields,
+                                                                   BType type) {
+        type = types.getReferredType(type);
+        if (type.tag != TypeTags.RECORD) {
+            return false;
+        }
+        BRecordType recordType = (BRecordType) type;
+        for (BLangRecordVarRefKeyValue lhsField : recordRefFields) {
+            if (recordType.fields.containsKey(lhsField.variableName.value)) {
+                if (Symbols.isOptional(recordType.fields.get(lhsField.variableName.value).symbol)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    private boolean checkValidMappingBindingPatternForRecordVariable(List<BLangRecordVariable.
```

Comment:
Should we rename this to something like `validateBindingPatternForNonRequiredFields`, reflecting the actual check?

Reply:
Assuming we move the logging to this method.